# Data Preprocessing

In [27]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import rcParams
from sklearn.impute import SimpleImputer
from scipy import stats

In [2]:
df = pd.read_excel('E Commerce Dataset.xlsx', sheet_name='E Comm')

In [3]:
df.drop('CustomerID', axis = 1, inplace = True)

In [5]:
df.sample(5)

Churn  Tenure PreferredLoginDevice  CityTier  WarehouseToHome  \
4517      1     1.0             Computer         3              7.0   
2289      0     1.0             Computer         1              7.0   
1525      0     4.0                Phone         1             18.0   
682       0    11.0         Mobile Phone         3             16.0   
3400      0     7.0             Computer         3             24.0   

     PreferredPaymentMode  Gender  HourSpendOnApp  NumberOfDeviceRegistered  \
4517                  COD  Female             4.0                         5   
2289           Debit Card  Female             2.0                         5   
1525          Credit Card  Female             3.0                         3   
682              E wallet    Male             3.0                         4   
3400           Debit Card    Male             3.0                         4   

        PreferedOrderCat  SatisfactionScore MaritalStatus  NumberOfAddress  \
4517        Mobile Phone                  4        Single                3   
2289  Laptop & Accessory                  1       Married                1   
1525        Mobile Phone                  1       Married                9   
682              Fashion                  3       Married                1   
3400             Grocery                  3        Single                3   

      Complain  OrderAmountHikeFromlastYear  CouponUsed  OrderCount  \
4517         1                         12.0         2.0         2.0   
2289         0                         11.0         7.0         7.0   
1525         1                         23.0         1.0         1.0   
682          1                         14.0         2.0         2.0   
3400         0                          NaN         6.0         7.0   

      DaySinceLastOrder  CashbackAmount  
4517                3.0          146.74  
2289                5.0          166.74  
1525                2.0          139.29  
682                 7.0          187.40  
3400                1.0          299.44

In [12]:
#pengelompokan
cats = ['Churn','CityTier','PreferredLoginDevice','PreferredPaymentMode','Gender','PreferedOrderCat','MaritalStatus']
nums = ['Tenure','WarehouseToHome','HourSpendOnApp','NumberOfDeviceRegistered','SatisfactionScore','NumberOfAddress'
            ,'Complain','OrderAmountHikeFromlastYear','CouponUsed','OrderCount','DaySinceLastOrder','CashbackAmount']

In [18]:
df['Churn'] = df['Churn'].astype(str)
df['CityTier'] = df['CityTier'].astype(str)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 19 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Churn                        5630 non-null   object 
 1   Tenure                       5366 non-null   float64
 2   PreferredLoginDevice         5630 non-null   object 
 3   CityTier                     5630 non-null   object 
 4   WarehouseToHome              5379 non-null   float64
 5   PreferredPaymentMode         5630 non-null   object 
 6   Gender                       5630 non-null   object 
 7   HourSpendOnApp               5375 non-null   float64
 8   NumberOfDeviceRegistered     5630 non-null   int64  
 9   PreferedOrderCat             5630 non-null   object 
 10  SatisfactionScore            5630 non-null   int64  
 11  MaritalStatus                5630 non-null   object 
 12  NumberOfAddress              5630 non-null   int64  
 13  Complain          

## Merge Value Ganda

In [8]:
#Sebelum Merge
for col in cats:
    print(f'''Value count kolom {col}:''')
    print(df[col].value_counts())
    print()

Value count kolom PreferredLoginDevice:
Mobile Phone    2765
Computer        1634
Phone           1231
Name: PreferredLoginDevice, dtype: int64

Value count kolom PreferredPaymentMode:
Debit Card          2314
Credit Card         1501
E wallet             614
UPI                  414
COD                  365
CC                   273
Cash on Delivery     149
Name: PreferredPaymentMode, dtype: int64

Value count kolom Gender:
Male      3384
Female    2246
Name: Gender, dtype: int64

Value count kolom PreferedOrderCat:
Laptop & Accessory    2050
Mobile Phone          1271
Fashion                826
Mobile                 809
Grocery                410
Others                 264
Name: PreferedOrderCat, dtype: int64

Value count kolom MaritalStatus:
Married     2986
Single      1796
Divorced     848
Name: MaritalStatus, dtype: int64



In [9]:
df.loc[df['PreferredLoginDevice'] == 'Phone', 'PreferredLoginDevice' ] = 'Mobile Phone'
df.loc[df['PreferredPaymentMode'] == 'COD', 'PreferredPaymentMode' ] = 'Cash on Delivery'
df.loc[df['PreferredPaymentMode'] == 'CC', 'PreferredPaymentMode' ] = 'Credit Card'
df.loc[df['PreferedOrderCat'] == 'Mobile', 'PreferedOrderCat' ] = 'Mobile Phone'

In [20]:
#Sesudah Merge
for col in cats:
    print(f'''Value count kolom {col}:''')
    print(df[col].value_counts())
    print()

Value count kolom Churn:
0    4682
1     948
Name: Churn, dtype: int64

Value count kolom CityTier:
1    3666
3    1722
2     242
Name: CityTier, dtype: int64

Value count kolom PreferredLoginDevice:
Mobile Phone    3996
Computer        1634
Name: PreferredLoginDevice, dtype: int64

Value count kolom PreferredPaymentMode:
Debit Card          2314
Credit Card         1774
E wallet             614
Cash on Delivery     514
UPI                  414
Name: PreferredPaymentMode, dtype: int64

Value count kolom Gender:
Male      3384
Female    2246
Name: Gender, dtype: int64

Value count kolom PreferedOrderCat:
Mobile Phone          2080
Laptop & Accessory    2050
Fashion                826
Grocery                410
Others                 264
Name: PreferedOrderCat, dtype: int64

Value count kolom MaritalStatus:
Married     2986
Single      1796
Divorced     848
Name: MaritalStatus, dtype: int64



## Missing Value Treatment

In [5]:
df.isna().sum()

CustomerID                       0
Churn                            0
Tenure                         264
PreferredLoginDevice             0
CityTier                         0
WarehouseToHome                251
PreferredPaymentMode             0
Gender                           0
HourSpendOnApp                 255
NumberOfDeviceRegistered         0
PreferedOrderCat                 0
SatisfactionScore                0
MaritalStatus                    0
NumberOfAddress                  0
Complain                         0
OrderAmountHikeFromlastYear    265
CouponUsed                     256
OrderCount                     258
DaySinceLastOrder              307
CashbackAmount                   0
dtype: int64

In [6]:
sum_missing = df.isna().sum()
percent_missing = round(df.isnull().sum() * 100 / len(df),2)
missing_value_df = pd.DataFrame({'column_name': df.columns,
                                 'sum_missing': sum_missing,
                                 'percent_missing': percent_missing}).reset_index(drop=True)
missing_value_df

column_name  sum_missing  percent_missing
0                    CustomerID            0             0.00
1                         Churn            0             0.00
2                        Tenure          264             4.69
3          PreferredLoginDevice            0             0.00
4                      CityTier            0             0.00
5               WarehouseToHome          251             4.46
6          PreferredPaymentMode            0             0.00
7                        Gender            0             0.00
8                HourSpendOnApp          255             4.53
9      NumberOfDeviceRegistered            0             0.00
10             PreferedOrderCat            0             0.00
11            SatisfactionScore            0             0.00
12                MaritalStatus            0             0.00
13              NumberOfAddress            0             0.00
14                     Complain            0             0.00
15  OrderAmountHikeFromlastYear          265             4.71
16                   CouponUsed          256             4.55
17                   OrderCount          258             4.58
18            DaySinceLastOrder          307             5.45
19               CashbackAmount            0             0.00

In [22]:
print(f'The number of missing values: {df.isnull().sum().sum()}')
print(f'The number of rows with missing values: {df[df.isnull().any(axis=1)].shape[0]}')

The number of missing values: 1856
The number of rows with missing values: 1856


### Treatment apa yang tepat untuk missing Values
Total ada 1.856 missing values, dan juga ada 1.856 rows dengan missing values, artinya setiap missing values berada pada row yang berbeda, dan tidak ada rows yang memiliki lebih dari satu missing values.
<br>
Jadi jika kita drop semua rows dengan missing values, kita akan membuang 1856 rows atau 32,97 persen dari dataset dan akan menyebabkan data loss, sehingga treatment yang paling tepat adalah dengan imputation

In [28]:
imputer = SimpleImputer(strategy= 'median')      
imputer.fit(df[nums])
df[nums] = imputer.transform(df[nums])

In [29]:
df.isna().sum()

Churn                          0
Tenure                         0
PreferredLoginDevice           0
CityTier                       0
WarehouseToHome                0
PreferredPaymentMode           0
Gender                         0
HourSpendOnApp                 0
NumberOfDeviceRegistered       0
PreferedOrderCat               0
SatisfactionScore              0
MaritalStatus                  0
NumberOfAddress                0
Complain                       0
OrderAmountHikeFromlastYear    0
CouponUsed                     0
OrderCount                     0
DaySinceLastOrder              0
CashbackAmount                 0
dtype: int64

## Outliers Treatment

In [30]:
print(f'Jumlah baris sebelum memfilter outlier: {len(df)}')

filtered_entries = np.array([True] * len(df))

for col in nums:
    zscore = abs(stats.zscore(df[col])) # hitung absolute z-scorenya
    filtered_entries = (zscore < 3) & filtered_entries # keep yang kurang dari 3 absolute z-scorenya
    
df = df[filtered_entries] # filter, cuma ambil yang z-scorenya dibawah 3

print(f'Jumlah baris setelah memfilter outlier: {len(df)}')

Jumlah baris sebelum memfilter outlier: 5630
Jumlah baris setelah memfilter outlier: 5350


## Feature Transformation

## Feature Encoding

## Class Imbalance